### PR TITLE
Hrcpp 456/counters

### PR DESCRIPTION
**Pull request by rigazilla**
This is an implementation of counters for C++

https://issues.jboss.org/browse/HRCPP-456

### CODE REVIEW

File: include/infinispan/hotrod/CounterConfiguration.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+};
+
+class HR_EXTERN CounterConfiguration {
+public:
+
+    CounterConfiguration(long initialValue, long lowerBound, long upperBound, int concurrencyLevel,
+            CounterType type, Storage storage)
+    :
+            initialValue(initialValue),
+                    upperBound(upperBound),
```
Comment (alanfx) [inline include/infinispan/hotrod/CounterConfiguration.h]: Is the indentation weird here?

File: src/hotrod/api/CountersImpl.h [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    virtual long getValue();
+
+    void* addListener(const event::CounterListener& listener);
+
+    void removeListener(void* handler);
+
+    virtual ~BaseCounterImpl() {
+    }
+
+//   public <T extends CounterListener> Handle<T> addListener(T listener) {
```
Comment (alanfx) [inline src/hotrod/api/CountersImpl.h]: Can this be removed?

File: src/hotrod/api/CountersImpl.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    //    /**
+    //     * Registers a {@link CounterListener} to this counter.
+    //     *
+    //     * @param listener The listener to register.
+    //     * @param <T>      The concrete type of the listener. It must implement {@link CounterListener}.
+    //     * @return A {@link Handle} that allows to remove the listener via {@link Handle#remove()}.
+    //     */
+    //    <T extends CounterListener> Handle<T> addListener(T listener);
```
Comment (alanfx) [inline src/hotrod/api/CountersImpl.h]: Same here?

File: src/hotrod/api/CountersImpl.h [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    std::string getName() {
+        return name;
+    }
+    CounterConfiguration getConfiguration() {
+        return configuration;
+    }
+
+    void setRemoved()
+    {
```
Comment (alanfx) [inline src/hotrod/api/CountersImpl.h]: The placement of the `{` is inconsistent. In some places it is on the same line as the method declaration, and in some places it is alone on a line. (I personally prefer the first!)

File: src/hotrod/api/RemoteCounterManagerImpl.cpp [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+void RemoteCounterManagerImpl::stop() {
+    started = false;
+    codec = nullptr;
+    transportFactory.reset();
+}
+
+std::shared_ptr<StrongCounter> RemoteCounterManagerImpl::getStrongCounter(std::string name) {
+    auto it = counters.find(name);
+    if (it == counters.end())
+            {
```
Comment (alanfx) [inline src/hotrod/api/RemoteCounterManagerImpl.cpp]: More inconsistent `{` placement in this file

File: src/hotrod/impl/event/ClientListenerNotifier.h [hunk 6]
```diff
@@ -25,9 +25,10 @@ class ClientListenerNotifier {
 public:
 	virtual ~ClientListenerNotifier();
 	void addClientListener(const std::vector<char> listenerId, const ClientListener& clientListener, const std::vector<char> cacheName,  Transport& t, const Codec20& codec20, std::shared_ptr<void> operationPtr, const std::function<void()> &recoveryCallback);
+	std::shared_ptr<CounterDispatcher> addCounterListener(const std::vector<char> listenerId, const std::vector<char> cacheName,  Transport& t, const Codec20& codec20, const std::function<void()> &recoveryCallback);
 	void removeClientListener(const std::vector<char> listenerId);
 	void startClientListener(const std::vector<char> listenerId);
-	const ClientListener& findClientListener(const std::vector<char> listenerId);
+//	const ClientListener& findClientListener(const std::vector<char> listenerId);
```
Comment (alanfx) [inline src/hotrod/impl/event/ClientListenerNotifier.h]: Remove?

File: xunit-test/CountersTest/StrongCounterTest.cpp [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    auto& counterManager = StrongCounterTest::remoteCacheManager->getCounterManager();
+    auto cc = CounterConfiguration(initialValue, -100, 200, 8, CounterType::UNBOUNDED_STRONG, Storage::VOLATILE);
+    counterManager.defineCounter(counterName, cc);
+    auto counter = counterManager.getStrongCounter(counterName);
+    EXPECT_EQ(counter->addAndGet(100), 105);
+    counter->remove();
+    EXPECT_EQ(counter->getValue(), 5);
+    EXPECT_EQ(counter->addAndGet(-100), -95);
+    counterManager.remove(counterName);
+    EXPECT_EQ(counter->getValue(), 5);
```
Comment (alanfx) [inline xunit-test/CountersTest/StrongCounterTest.cpp]: Shouldn't this usage of the counter after calling `CounterManager.remove(String)` cause an error?

File: xunit-test/CountersTest/WeakCounterTest.cpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    counterManager.defineCounter(counterName, cc);
+    auto counter = counterManager.getWeakCounter(counterName);
+    counter->add(100);
+    EXPECT_EQ(counter->getValue(),105);
+    counter->remove();
+    EXPECT_EQ(counter->getValue(),5);
+    counter->add(-100);
+    EXPECT_EQ(counter->getValue(),-95);
+    counterManager.remove(counterName);
+    EXPECT_EQ(counter->getValue(),5);
```
Comment (alanfx) [inline xunit-test/CountersTest/WeakCounterTest.cpp]: Same question here as above?

File: xunit-test/CountersTest/StrongCounterTest.cpp (hunk 7 above)
Comment (rigazilla) [reply xunit-test/CountersTest/StrongCounterTest.cpp]: No that's ok: the counter is recreated if accessed again. Here's the doc:

http://infinispan.org/docs/dev/user_guide/user_guide.html#hot_rod_protocol_2_7

**Comment by rigazilla** [comment]
fixed formatting

File: xunit-test/CountersTest/StrongCounterTest.cpp (hunk 7 above)
Comment (alanfx) [reply xunit-test/CountersTest/StrongCounterTest.cpp]: OK, I see now. I think you mean this link:

[http://infinispan.org/docs/dev/user_guide/user_guide.html#remove_a_counter_via_countermanager](http://infinispan.org/docs/dev/user_guide/user_guide.html#remove_a_counter_via_countermanager)

Comment (rigazilla) [reply xunit-test/CountersTest/StrongCounterTest.cpp]: Yeah, but this is for the embedded usage, the one above is the for romote via hotrod (see Counter remove operation parag).

**Comment by alanfx** [comment]
@rigazilla There are still a couple of lonely `{` on a line by themselves, but I'll go ahead and merge this!

**Pull request merged**